aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/io/MutableBufferNative.kt

```diff
@@ -4,33 +4,65 @@
  */
 package aws.sdk.kotlin.crt.io
 
+import aws.sdk.kotlin.crt.Allocator
+import aws.sdk.kotlin.crt.Closeable
+import kotlinx.cinterop.*
+import libcrt.*
+
 /**
  * Represents a mutable linear range of bytes that can be written to.
  * Instance of this class has no additional state except the bytes themselves.
- *
- * NOTE: Platform implementations should provide direct access to the underlying bytes
  */
-public actual class MutableBuffer {
+@OptIn(ExperimentalForeignApi::class)
+public actual class MutableBuffer(buffer: aws_byte_buf? = null, private val capacity: Int) : Closeable { // TODO implement CrtResource?
+    private val buf = buffer ?: Allocator.Default.alloc<aws_byte_buf>()
+
+    public val bytes: ByteArray
+        get() = buf.buffer!!.readBytes(buf.capacity.toInt())
+
+    init {
+        if (buffer == null) {
+            aws_byte_buf_init(buf = buf.ptr, allocator = Allocator.Default.allocator, capacity = capacity.toULong())
+        }
+    }
+
     /**
      * The amount of remaining write capacity before the buffer is full
      */
     public actual val writeRemaining: Int
-        get() = TODO("Not yet implemented")
+        get() = buf.capacity.toInt() - buf.len.toInt()
 
     /**
      * Write as much of [length] bytes from [src] as possible starting at [offset].
      * The number of bytes written is returned which may be less than [length]
      */
     public actual fun write(src: ByteArray, offset: Int, length: Int): Int {
-        TODO("Not yet implemented")
+        src.usePinned { pinnedSrc ->
+            val offsetPinnedSrc = pinnedSrc.addressOf(offset).reinterpret<UByteVar>()
+            if (aws_byte_buf_write(buf.ptr, offsetPinnedSrc, length.toULong())) {
```

Comment:
**correctness**: This will only write if there is enough size for `length` bytes. I think we want to calculate the `minOf(length, writeRemaining)` and write as much as we can to the buffer.